CsharpExtras/ValidatedType/Validated.cs

```diff
@@ -6,26 +6,32 @@ namespace CsharpExtras.ValidatedType
 {
     public abstract class Validated<TVal>
     {
-        private readonly TVal Value;
+        private readonly TVal _value;
         protected Validated(TVal value)
         {
             if (IsValid(value))
             {
-                Value = value;
+                _value = value;
             }
             else
             {
-                string valueAsString = value?.ToString() ?? "NULL";
+                string valueAsString = GetValueAsString(value);

```

Comment:
I may be nitpicking, but if the value is immutable, shouldn't the string value be immutable too? In that case it can be calculated just once and stored on the class as a variable.

Reply:
Good point. We could create an immutable supertype that lazily stores the value for toString() and then delegates down to some abstract function to do the first calc

Reply:
Or maybe an immutable supertype is overkill... I'll just make it lazy here

Reply:
After coding up the change, twice, I realise that this isn't a good approach, in general. Even though the reference is immutable, doesn't mean the object data is. So `toString()` could return something different upon two successive calls from the same object. I attempted to rectify this by enforcing the generic parameter to be a struct, i.e. a value type, but this doesn't solve the above issue and it also disallows string, which is bad. So I think the best approach is to leave it as is, it's pretty cheap to call toString() anyway, and I'm sure the underlying value types, if they are indeed immutable, will have some optimisation there anyway, similar to what you suggest.

Reply:
> Even though the reference is immutable, doesn't mean the object data is

Doesn't this undermine the purpose of the validator?

If the data you are validating can change, it isn't enough to only validate it on construction because the value might change to something invalid.

Reply:
> Doesn't this undermine the purpose of the validator?

Good point. Sub-classes of this abstract class should only use an immutable type as the generic parameter. That would be either a primitive value type or something like string. There's no way to enforce that with generics though.

Stepping back out to the original comment, which was the efficiency of `toString()`, the responsibility to make that efficient is on the owner of the immutable type itself. So this class shouldn't take that responsibility and get confused about its purpose.

Reply:
Yeah, that all makes sense. Ideally the validator class would be able to enforce that the types being validated are immutable, but it looks like there is no way to do it. In that case I think just documenting that the validator class is built around that assumption is enough.